### ISSUE TITLE
ZQL needs to throw better errors

### ISSUE THREAD

**Issue by adityagp** [zenvisage/zenvisage#164]
Right now, there are no error messages provided by the ZQL execution engine; it just silently fails. 

This is important feedback for users who want to use it.

**Comment by adityagp** [comment]
I feel like a simple error message that says "Malformed ZQL" is a valuable start; even something as simple as "Unknown variable 'X1'" or "Uninterpretable expression 'State.'CA'" would be very useful

**Comment by siddta** [comment]
We need to update the Results class with two additional attributes: isSuccess and errorMessage. The back-end should set isSuccess to false if something goes wrong; and should map the actual error to a more understandable message based on the component of zv that is throwing the error. So we should have a list of error-messages (user-defined exceptions) stored somewhere.
Front-end before parsing the results should check isSuccess. If it's set to false, it should display the errorMessage. 
@adityagp I think we need to have a more generic error propagation framework, so it's not just limited to ZQL. I think I have another issue opened for this. Perhaps, we might not be able to finish this until 2.0 release, since it would touch almost all parts of zv, but it's an important change.

**Comment by adityagp** [comment]
Agreed that this is more involved, but right now, is there no way for the backend to communicate that there is an error in ZQL? For example, it could set the result list to be a nullset.

**Comment by siddta** [comment]
right now I think it either throws an actual java error message, or doesn't return anything, both of which front-end fails to understand. Adding isSuccess and errorMessage to the output should fix.

**Comment by zwang180** [comment]
Not sure if this is a good idea, but you could let the backend return the current result class along with setting the http status code to 500, which indicate internal server error, when something is wrong in the backend. Thus the frontend can just add a error handling function like this(use my regex execution code as an example):

```
        $http.post('/zv/executeRegexShapeQuery', data).
        success(function(response) {
            // If success
            console.log("execute Regex: success");
            if (response.length == 0){console.log("empty response")}
            console.log(response);
            plotResults.displayUserQueryResults(response.outputCharts,$scope.flipY,true);
        }).
        error(function(response) {
            // If error, more error handling logic
            console.log("getUserQueryResults: fail");
        });
```
For the future, add error message field to result class, so the front end can just grab it and inform the user the exact error in the backend.

**Comment by siddta** [comment]
that's a good idea @zwang180, instead of status code we can maybe set isSuccess variable to false, but either way is fine.

**Comment by eidehua** [comment]
I do think responding with status codes would be the better way to do it. Normal execution would just return 200 OK, if we get a backend exception we can return 500

**Comment by siddta** [comment]
okay, we can go with status=500 instead of isSuccess=false, if it doesn't break the front-end. The other thing is that we should catch all the exceptions (we are currently not catching quite a few of them, resulting in backend failing silently), and then convert those exceptions to meaningful messages.

**Comment by eidehua** [comment]
Confirmed we can set status to 500, using 
response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, "hey");

![image](https://user-images.githubusercontent.com/7684873/30903388-39794c2c-a334-11e7-9a34-9ed88d18104a.png)

Now frontend just needs to handle errors

**Comment by adityagp** [comment]
Looks good to me! For now, can we just throw an error message at the front end, without worrying about it too much?

**Comment by eidehua** [comment]
Yea, I can add some basic error handling on frontend

**Comment by eidehua** [comment]
Alright, on the front end when it receives an error, I created a nice modal popup to contain the error.

![image](https://user-images.githubusercontent.com/7684873/31151620-e415a06c-a85d-11e7-9202-3b682e105a78.png)

The frontend actually already catches errors, but just silently drops them (well just printing to console).

**Comment by adityagp** [comment]
Excellent, thanks Ed!

**Comment by eidehua** [comment]
Backend Implementation:

![image](https://user-images.githubusercontent.com/7684873/31162247-f120ca9a-a8a1-11e7-9125-3b9093cb5bd4.png)

From the frontend, it may be weird to see extensive java error messages. So instead of throwing up the JSONParseException, I make a simple message for the frontend to see. 

The benefit of just sending the real exception is so for easier debugging. If we want to show real exception messages I can do that too.

**Comment by eidehua** [comment]
The cost of this however is a very ugly postSimilarity Function

![image](https://user-images.githubusercontent.com/7684873/31162426-03ab5cd8-a8a3-11e7-946d-82f4b8bede65.png)

If the scanner fails, or if removingSketchPoints fails, or if the main algorithm fails, we want to return an error message to front end, so several catch/try blocks

The nested catching occurs because there may be an IOException when we send something to the frontend! (client closed connection). We can remove these nested catches by adding throws IOException-- but this can hide some real IOExceptions we want to sendErrors for (say alter we add some functionality, and that throws an IOException-- compiler will be fine with it since our function throws IOException)

**Comment by eidehua** [comment]
OK I think I found a better solution, with just one big try catch block. Let the called functions return me either their standard error messages, or simplified ones.

![image](https://user-images.githubusercontent.com/7684873/31162781-e58c3996-a8a4-11e7-9f0f-581af4064490.png)

**Comment by eidehua** [comment]
Implemented the changes we discussed in last week's meeting.

![image](https://user-images.githubusercontent.com/7684873/31416618-f108d464-adef-11e7-9cab-64206f37523e.png)

![image](https://user-images.githubusercontent.com/7684873/31416624-fb001c66-adef-11e7-8a49-26742df4d11e.png)

**Comment by eidehua** [comment]
Closed with 1f6c7173b2b67e273af8c7bed8e33a77296e193e
